### ISSUE TITLE
v3.0.0-beta.1 - panic: redis configuration required to use for layerinfo cache

### ISSUE THREAD

**Issue by Vad1mo** [distribution/distribution#4486]
### Description

I am currently testing docker distribution beta for project Harbor. And it seems like our old redis config does not work with distribution anymore.  

```console
anic: redis configuration required to use for layerinfo cache

goroutine 1 [running]:
github.com/distribution/distribution/v3/registry/handlers.NewApp({0x5613b94d5528, 0xc000741380}, 0xc0006c9208)
	/go/src/github.com/docker/distribution/registry/handlers/app.go:288 +0x1ebd
github.com/distribution/distribution/v3/registry.NewRegistry({0x5613b94d5528?, 0xc0007405d0?}, 0xc0006c9208)
	/go/src/github.com/docker/distribution/registry/registry.go:142 +0x56
github.com/distribution/distribution/v3/registry.init.func1(0x5613b9ffa800, {0xc00074e0d0, 0x1, 0x1})
	/go/src/github.com/docker/distribution/registry/registry.go:111 +0x10a
github.com/spf13/cobra.(*Command).execute(0x5613b9ffa800, {0xc00074e090, 0x1, 0x1})
	/go/src/github.com/docker/distribution/vendor/github.com/spf13/cobra/command.go:987 +0xab1
github.com/spf13/cobra.(*Command).ExecuteC(0x5613b9ffaae0)
	/go/src/github.com/docker/distribution/vendor/github.com/spf13/cobra/command.go:1115 +0x3ff
github.com/spf13/cobra.(*Command).Execute(0xc0000061c0?)
	/go/src/github.com/docker/distribution/vendor/github.com/spf13/cobra/command.go:1039 +0x13
main.main()
	/go/src/github.com/docker/distribution/cmd/registry/main.go:26 +0x1a

```

### Reproduce

```yaml
version: 0.1
log:
  level: info
  fields:
    service: registry
storage:
  s3:
    region: xxx
    bucket: xxxx
    regionendpoint: xxx
  cache:
    layerinfo: inmemory
  maintenance:
    uploadpurging:
      enabled: true
      age: 168h
      interval: 24h
      dryrun: false
  delete:
    enabled: true
  redirect:
    disable: false
redis:
  addr: redis-master:6379
  db: 2
  readtimeout: 10s
  writetimeout: 10s
  dialtimeout: 10s
  pool:
    maxidle: 100
    maxactive: 500
    idletimeout: 60s
http:
  addr: :5000
  relativeurls: false
  # set via environment variable
  # secret: placeholder
  debug:
    addr: :8001
    prometheus:
      enabled: true
      path: /metrics
auth:
  htpasswd:
    realm: harbor-registry-basic-realm
    path: /etc/registry/passwd
validation:
  disabled: true
compatibility:
  schema1:
    enabled: true

```

### Expected behavior

The same configuration that worked for 2.8.x also works for 3.0.x

### registry version

v3.0.0-beta.1

**Comment by milosgajdos** [comment]
Note that `layerinfo` has been deprecated and replaced by `blobdescriptor`: 
https://distribution.github.io/distribution/about/configuration/#cache

**Comment by Vad1mo** [comment]
> Note that `layerinfo` has been deprecated and replaced by `blobdescriptor`: https://distribution.github.io/distribution/about/configuration/#cache

@milosgajdos thank you for that tip, I have noticed that change and also tried out changing the config to  blobdescriptor: redis, unfortunaltly with the same result.

**Comment by milosgajdos** [comment]
Very strange! We use this config in our integration tests which run on every PR: https://github.com/distribution/distribution/blob/d0eebf3af4fc1d5c0287e5af61147403ccb78ec2/tests/conf-e2e-cloud-storage.yml#L19-L32

I will have to have a proper look at this; my initial suspicion is some YAML shenanigans or some such, but let's not draw any conclusions, yet!

**Comment by milosgajdos** [comment]
Actually, it's the well-known silly `configuration` package in this repo that is just so terrible that a part of me dies everytime I open it.

TL;DR: 
Please change this config
```yaml
redis:
  addr: redis-master:6379
```

to
```yaml
redis:
  addrs: [localhost:6379]
```

Here's the link to the docs:
https://distribution.github.io/distribution/about/configuration/#redis

The `configuration` package has been a source of too many woes (i.e. it's not helpful when it barfs its guts) over the years and would need a rewrite from the ground up, unfortunately.

**Comment by Vad1mo** [comment]
What a nightmare for operators, So the missing trick is the `s` suffix in addr: --> addr**s**:. The change to array format  address is obvious. But this is clear in the docs, not so the `s`.

Maybe to not annoy too many people and break migrations, docker distribution could support the old format `addr: redis-master:6379` too.

**Comment by milosgajdos** [comment]
> Maybe to not annoy too many people and avoid this silly breaking change, docker distribution could support the old format addr: redis-master:6379 too.

This would require some changes that would have made the already appallingly awful `configuration` package even worse and likely given me a mild stroke, so I'd rather....not

**Comment by milosgajdos** [comment]
@Vad1mo mind closing this issue?

**Comment by shanduur** [comment]
> The `configuration` package has been a source of too many woes (i.e. it's not helpful when it barfs its guts) over the years and would need a rewrite from the ground up, unfortunately.

Maybe it's time to do that? Before v3 goes live?

**Comment by milosgajdos** [comment]
> Maybe it's time to do that? Before v3 goes live?

We've discussed that a few times but that'd be a rather chunky undertaking and given we are very far down the v3 release path, it would delay the whole process further still (release, test, another release, etc.).

**Comment by Vad1mo** [comment]
its going to be a breaking change from v2 to v3, so far this was the only config I found. It took me half a day to detect, because it is so subtle.

**Comment by milosgajdos** [comment]
> its going to be a breaking change from v2 to v3, so far this was the only config I found. It took me half a day to detect, because it is so subtle.

major releases allow for breaking changes; in fact that should generally be expected from major releases

**Comment by laur89** [comment]
According to [docs](https://distribution.github.io/distribution/about/configuration/#redis), following env var should work: `REGISTRY_REDIS_ADDRS_0=redis:port`, yet when v3 is started up, it fails with following logs:
```
time="2025-04-24T16:27:14Z" level=warning msg="Ignoring unrecognized environment variable REGISTRY_REDIS_ADDRS_0"
time="2025-04-24T16:27:14.42729384Z" level=debug msg="using \"text\" logging formatter"
time="2025-04-24T16:27:14.428120255Z" level=info msg="redis not configured" environment=development go.version=go1.23.7 instance.id=86a40e80-8f98-467d-8325-0a2309091234 service=registry version=3.0.0
time="2025-04-24T16:27:14.428226011Z" level=info msg="Starting upload purge in 29m0s" environment=development go.version=go1.23.7 instance.id=86a40e80-8f98-467d-8325-0a2309091234 service=registry version=3.0.0
panic: redis configuration required to use for layerinfo cache

goroutine 1 [running]:
github.com/distribution/distribution/v3/registry/handlers.NewApp({0x21c7c80, 0xc0001228d0}, 0xc000887408)
        github.com/distribution/distribution/v3/registry/handlers/app.go:288 +0x1edd
github.com/distribution/distribution/v3/registry.NewRegistry({0x21c7c80?, 0xc00078b920?}, 0xc000887408)
        github.com/distribution/distribution/v3/registry/registry.go:150 +0x56
github.com/distribution/distribution/v3/registry.init.func1(0x32fdbc0, {0xc0007a2950, 0x1, 0x1})
        github.com/distribution/distribution/v3/registry/registry.go:119 +0x10a
github.com/spf13/cobra.(*Command).execute(0x32fdbc0, {0xc0007a2910, 0x1, 0x1})
        github.com/spf13/cobra@v1.8.0/command.go:987 +0xa91
github.com/spf13/cobra.(*Command).ExecuteC(0x32fdea0)
        github.com/spf13/cobra@v1.8.0/command.go:1115 +0x3ff
github.com/spf13/cobra.(*Command).Execute(0xc0000061c0?)
        github.com/spf13/cobra@v1.8.0/command.go:1039 +0x13
main.main()
        github.com/distribution/distribution/v3/cmd/registry/main.go:26 +0x1a
```


```
$ docker run
  -d
  --name='docker-registry'
  -e 'REGISTRY_STORAGE_DELETE_ENABLED'='true'
  -e 'REGISTRY_HTTP_SECRET'='secret'
  -e 'REGISTRY_STORAGE_CACHE_BLOBDESCRIPTOR'='redis'
  -e 'REGISTRY_REDIS_ADDRS_0'='my-redis-host:6379'
  -p '5000:5000/tcp'
  -v '/path/on/host/registry':'/var/lib/registry':'rw'
  'registry:3' 
```

**Comment by thaJeztah** [comment]
Does it work if you use `-e 'REGISTRY_REDIS_ADDRS'='[my-redis-host:6379]'` ? See https://github.com/distribution/distribution/issues/4621#issuecomment-2826218587

**Comment by laur89** [comment]
Nope: `level=warning msg="Ignoring unrecognized environment variable REGISTRY_REDIS_ADDRS"`

**Comment by thaJeztah** [comment]
Ah, sorry meant to write `REGISTRY_REDIS_OPTIONS_ADDRS` (see the linked comment)
